odoo/addons/base/ir/ir_ui_view.py

```diff
@@ -853,6 +869,60 @@ def collect(node, model):
 
         return arch
 
+    def get_attrs_symbols(self):
+        """ Return a list of predefined symbols for evaluating attrs. """
+        return ['self', 'parent', 'id', 'uid', 'context',
+                'active_id', 'active_ids', 'active_model',
+                'time', 'datetime', 'relativedelta',
+                'context_today', 'current_date']
+
+    def get_attrs_field_names(self, arch):
+        """ Retrieve the field names appearing in context, domain and attrs, and
+            return a list of triples ``(field_name, attr_name, attr_value)``.
+        """
+
+        def get_name(node):
+            """ return the name from an AST node, or None """
+            if isinstance(node, ast.Name):
+                return node.id
+
+        def get_parent(get, node):
+            """ return the subfield name from an AST node, or None """
+            if isinstance(node, ast.Attribute) and get(node.value) == 'parent':
+                return node.attr
+
+        def expr_names(expr, get):
+            """ parse `expr` and return field names """
+            return (name
+                    for node in ast.walk(ast.parse(expr.strip(), mode='eval'))
+                    for name in [get(node)]
+                    if name)
+
+        def node_names(node, get=get_name):
+            """ traverse `node` and return triples """
+            for key in ATTRS_WITH_FIELD_NAMES:
+                value = node.get(key)
+                if value:
+                    for name in expr_names(value, get):
+                        yield (name, key, value)
+            if node.get('attrs'):
+                attrs = node.get('attrs')
+                # collect field names in left-hand-side of conditions
+                for domain in safe_eval(attrs).values():
+                    if isinstance(domain, list):
```

Comment:
Is there any case where the domain could be a non-list?

Reply:
It once crashed with a boolean domain. Maybe we should fix the domain...